File: app/src/androidTest/java/TestGridRandomizer.java
import com.holokenmod.Grid;
import com.holokenmod.GridSize;
import com.holokenmod.creation.GridRandomizer;
import com.holokenmod.options.DigitSetting;
import com.holokenmod.options.GameVariant;

import org.junit.jupiter.api.Test;

public class TestGridRandomizer {
	@Test
	void test() {
		GameVariant.getInstance().setDigitSetting(DigitSetting.FIRST_DIGIT_ONE);
		GameVariant.getInstance().setShowOperators(true);
		
		Grid grid = new Grid(new GridSize(6, 3));
		
		grid.addAllCells();
		
		System.out.println(grid.toString());
		
		grid.clearUserValues();
		GridRandomizer randomizer = new GridRandomizer(grid);
		
		randomizer.createGrid();
		
		System.out.println(grid.toString());
	}
}

File: gauguin-core/src/test/kotlin/org/piepmeyer/gauguin/creation/TestGridRandomizer.kt
package org.piepmeyer.gauguin.creation

import io.kotest.assertions.withClue
import io.kotest.common.ExperimentalKotest
import io.kotest.core.spec.style.FunSpec
import io.kotest.datatest.withData
import io.kotest.matchers.collections.shouldBeIn
import io.kotest.matchers.shouldBe
import org.piepmeyer.gauguin.grid.Grid
import org.piepmeyer.gauguin.grid.GridSize
import org.piepmeyer.gauguin.options.GameOptionsVariant.Companion.createClassic
import org.piepmeyer.gauguin.options.GameVariant
import kotlin.time.Duration.Companion.seconds

@OptIn(ExperimentalKotest::class)
class TestGridRandomizer :
    FunSpec({
        context("digitsFromOneOn").config(timeout = 10.seconds) {

            val parameters = mutableListOf<Pair<Int, Int>>()

            for (width in 3..11) {
                for (height in 3..11) {
                    parameters += Pair(width, height)
                }
            }

            withData(
                parameters,
            ) { (width, heigth) ->
                run {
                    val grid = Grid(GameVariant(GridSize(width, heigth), createClassic()))
                    grid.clearUserValues()
                    val randomizer = GridRandomizer(RandomPossibleDigitsShuffler(), grid)
                    randomizer.createGridValues()
                    for (cell in grid.cells) {
                        withClue("Invalid solution of $grid") {
                            cell.value shouldBeIn grid.variant.possibleDigits
                        }
                    }
                }
            }
        }

        test("deterministic random number generator leads to deterministic grids") {
            val variant = GameVariant(GridSize(9, 9), createClassic())

            val gridOne = Grid(variant)
            val gridTwo = Grid(variant)

            randomizeGrid(gridOne)
            randomizeGrid(gridTwo)

            gridOne.toString() shouldBe gridTwo.toString()
        }
    })

private fun randomizeGrid(grid: Grid) {
    val random = SeedRandomizerMock(1)
    val randomizer = GridRandomizer(RandomPossibleDigitsShuffler(random.random), grid)
    randomizer.createGridValues()
}
